### PR TITLE
Add python3 link to gdb

### PR DESCRIPTION
**Pull request by Congee**
1st, removed building error caused by disutils
``` python
﻿﻿>>> from distutils import sysconfig
>>> getvar = sysconfig.get_config_var
>>> getvar('LINKFORSHARED')
'-u _PyMac_Error /usr/local/opt/python3/Frameworks/Python.framework/Versions/3.5/Python'
```
```
 Undefined symbols for architecture x86_64:
   "_PyMac_Error", referenced from:
      -u command line option
 ld: symbol(s) not found for architecture x86_64
 clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

2nd, I have to remove the option `--with-system-readline` to build with `python3`. But gdb actually is linked aginst with system readline (you don't have to `brew install readline`). Any suggestions?

### CODE REVIEW

File: Library/Formula/gdb.rb [hunk 1]
```diff
@@ -37,22 +51,29 @@ class Gdb < Formula
     depends_on UniversalBrewedPython
   end
 
+  if build.with? "brewed-python3"
+    depends_on UniversalBrewedPython3
+		patch :p1, :DATA
```
Comment (mikemcquaid) [inline Library/Formula/gdb.rb]: Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

**Comment by mikemcquaid** [comment]
What's the advantage of using Python 3, here?

File: Library/Formula/gdb.rb (hunk 1 above)
Comment (Congee) [inline Library/Formula/gdb.rb]: Thanks for your guidance; I'm reporting this to gdb bug tracker.

**Comment by Congee** [comment]
Because some people need to debug their programs which use Python 3 C api. :D

**Comment by mikemcquaid** [comment]
> Because some people need to debug their programs which use Python 3 C api. :D

@Congee and that doesn't work unless it's actually linked against Python 3?

File: Library/Formula/gdb.rb [hunk 2]
```diff
@@ -37,22 +51,29 @@ class Gdb < Formula
...
-    if build.with? "brewed-python"
-      args << "--with-python=#{HOMEBREW_PREFIX}"
-    else
-      args << "--with-python=/usr"
+    if build.with? "brewed-python" 
+      args << "--with-python=#{HOMEBREW_PREFIX}/bin/python"
+		elsif build.with? "brewed-python3"
+      args << "--with-python=#{HOMEBREW_PREFIX}/bin/python3"
+		else
+      args << "--with-python=/usr/local"
```
Comment (DomT4) [inline Library/Formula/gdb.rb]: `/usr`, not `/usr/local` here.

**Comment by Congee** [comment]
@mikemcquaid Sorry, my mistake. This pull request just enables gdb to execute Python 3 scirpt. It has nothing to do with debuging with Python 3 C api.

**Comment by Congee** [comment]
This is an upstream issue that gdb building scirpt uses an outdated python-config.py file, which has been fixed in official repo 2 years ago.

For more information, see:
gdb bug tracker: https://sourceware.org/bugzilla/show_bug.cgi?id=19378
python official patch: https://hg.python.org/cpython/rev/bab708624dc4/

File: Library/Formula/gdb.rb (hunk 2 above)
Comment (Congee) [inline Library/Formula/gdb.rb]: FYI, fixed. Thanks.

File: Library/Formula/gdb.rb [hunk 3]
```diff
@@ -37,20 +51,27 @@ class Gdb < Formula
     depends_on UniversalBrewedPython
   end
 
+  if build.with? "brewed-python3"
+    depends_on UniversalBrewedPython3
+    patch :p1, :DATA
```
Comment (mikemcquaid) [inline Library/Formula/gdb.rb]: Can you please submit this patch to the upstream developers of this project (if necessary) and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

File: Library/Formula/gdb.rb [hunk 4]
```diff
@@ -25,6 +38,7 @@ class Gdb < Formula
   end
 
   option "with-brewed-python", "Use the Homebrew version of Python"
+  option "with-brewed-python3", "Use the Homebrew version of Python3"
```
Comment (mikemcquaid) [inline Library/Formula/gdb.rb]: Make this option called `with-python3` instead

**Comment by Lingnik** [comment]
Hi! `brew install python3` helpfully pointed me to this PR as the potential resolution to the output I paste below.
• What is blocking the merge of this PR? Is there anything I can do to help?
• Is there a workaround?

```
cb-taylor:~ lingnik$ brew postinstall python3
==> /usr/local/Cellar/python3/3.4.3/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr/local/Cellar/python3/3.4.3/bin --install-lib=/usr/local/lib/python3.4/site-packages
  File "/usr/local/Cellar/python3/3.4.3/Frameworks/Python.framework/Versions/3.4/lib/python3.4/zipfile.py", line 904, in __init__
    _check_compression(compression)
  File "/usr/local/Cellar/python3/3.4.3/Frameworks/Python.framework/Versions/3.4/lib/python3.4/zipfile.py", line 585, in _check_compression
    "Compression requires the (missing) zlib module")
RuntimeError: Compression requires the (missing) zlib module

READ THIS: https://git.io/brew-troubleshooting

These open issues may also help:
Add python3 link to gdb (https://github.com/Homebrew/homebrew/pull/47146)
cb-taylor:~ lingnik$ python3 --version
Python 3.4.3
cb-taylor:~ lingnik$ which python3
/usr/local/bin/python3
cb-taylor:~ lingnik$ which brew
/usr/local/bin/brew
cb-taylor:~ lingnik$ brew --version
0.9.5
cb-taylor:~ lingnik$ brew update
fatal: unable to access 'https://github.com/Homebrew/homebrew/': Failed to connect to 172.20.33.70 port 8080: Operation timed out
Error: Failure while executing: git pull -q origin refs/heads/master:refs/remotes/origin/master
```

File: Library/Formula/gdb.rb [hunk 5]
```diff
@@ -11,6 +11,19 @@ def message; <<-EOS.undent
...
 
+class UniversalBrewedPython3 < Requirement
+  satisfy { archs_for_command("python3").universal? }
+
+  def message; <<-EOS.undent
+    A build of GDB using a brewed Python3 was requested, but Python3 is not
+    a universal build.
+
+    GDB requires Python3 to be built as a universal binary or it will fail
+    if attempting to debug a 32-bit binary on a 64-bit host.
```
Comment (mikemcquaid) [inline Library/Formula/gdb.rb]: This applies with normal GDB too so I don't think it should be a requirement but instead a caveat.

File: Library/Formula/gdb.rb [hunk 6]
```diff
@@ -39,11 +53,17 @@ class Gdb < Formula
     depends_on UniversalBrewedPython
   end
 
+  if build.with? "python3"
+    depends_on UniversalBrewedPython3
+    patch :p1, :DATA
```
Comment (mikemcquaid) [inline Library/Formula/gdb.rb]: Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

File: Library/Formula/gdb.rb [hunk 7]
```diff
@@ -39,11 +53,17 @@ class Gdb < Formula
...
+  end
+
   def install
     args = [
       "--prefix=#{prefix}",
       "--disable-debug",
       "--disable-dependency-tracking",
+      "--with-lzma",
```
Comment (mikemcquaid) [inline Library/Formula/gdb.rb]: Why is this needed? I don't object, just wondering.

**Comment by mikemcquaid** [comment]
@Lingnik I've made some comments here on what's blocking. Feel free to address them and make your PR as that may be quicker.

File: Library/Formula/gdb.rb (hunk 7 above)
Comment (Congee) [inline Library/Formula/gdb.rb]: Shall be deleted. The new CPython configure script has removed such option.

**Comment by Congee** [comment]
Yes, this patch looks ugly and should be resolved by upstream maintainers.
I reported this issue to gdb bug tracker. Nobody cared. The gdb gets built well on their platforms with old `python-config`.  Anyway, I tried.

@Lingnik If you need this patch, patch and build by yourself.

**Comment by MikeMcQuaid** [comment]
Passing on this as-is, please reopen on https://github.com/homebrew/homebrew-core if desired.
